### ISSUE TITLE
Platform-dependent currentExtent not clear enough

### ISSUE THREAD

**Issue by krOoze** [KhronosGroup/Vulkan-Docs#590]
There is some ambiguity in `currentExtent` descriptions in the platform-dependent surface documentation.

1) On some platforms there is something like:

> With Win32, `minImageExtent`, `maxImageExtent`, and `currentExtent` are the window size. Therefore, a swapchain’s `imageExtent` **must** match the window’s size.

Which sorta implies `minImageExtent == maxImageExtent == currentExtent`. But that is probably not true based on the developments in 1.0.63, and based on AMD implementation behavior I tested.

2) For Wayland it says:

> On Wayland, `currentExtent` is undefined (0xFFFFFFFF, 0xFFFFFFFF). Whatever the application sets a swapchain’s `imageExtent` to will be the size of the window, after the first image is presented.

I find it bit ambiguous. I am not sure if `currentExtent` is always `0xFFFFFFFF`, or if it gets properly updated to the window size after the first image is presented. And resp., if it is usable information for window resize events, or some platform-dependent API must be used instead.

3) On VI it says:
>The `currentExtent` of a VI surface is always undefined. 

Which makes it ambiguous whether it is undefined "undefined", or `0xFFFFFFFF` kind of undefined.

4) Should `minImageExtent` be allowed to be greater than `currentExtent`? Restriction on this seemed to have been removed in 1.0.63.

**Comment by jjulianoatnv** [comment]
Checking in...this issue was on the agenda this week for the working group, but we didn't get to it before running out of time. It will carry over to next week's agenda.

> Which sorta implies `minImageExtent == maxImageExtent == currentExtent`. 

Correct.

Win32 requiring swapchain extent to match currentExtent was already discussed within the working group several months ago. At the time it was decided to retain this restriction for win32 swapchains.  If this restriction were lifted, then it would become necessary to define how size-mismatch present operates. Options include:
- 1:1 centered
- 1:1 anchored in one of 8 directions
- stretched
- aspect-preserving stretch (and define what happens to "side bars")

On at least one Vulkan-supporting platform this behavior is specified as attributes of the native window object, and Vulkan on that platform inherits an answer from the native window object.

On at least one platform this behavior is not specified anywhere, but the platform disallowes size mismatch to occur, so it's a non-issue.

On Windows, this behavior is not an attribute of the native window object on which Vulkan is layered. This behavior does happen to be defined by DXGI for DXGI swapchains, but Vulkan's swapchain is not layered on DXGI swapchain, so Vulkan does not inherit an answer (from either DXGI or the native window object).

It might be interesting for a new Vulkan win32 extension to define this behavior for win32 swapchains, and to enable size-mismatched present. Such an extension does not exist presently.

**Comment by krOoze** [comment]
> > Which sorta implies minImageExtent == maxImageExtent == currentExtent.
> 
> Correct.

On AMD impl. it actually does not hold. I can get currentExtent == (0,0) and minImageExtent == (0,0).
Also in 1.0.63 restriction on `minImageExtent ` was removed in the non-platform-specific section AIS in the commit log comments.
Might it make sense to legitimize this behavior, or should AMD be notified?

**Comment by jjulianoatnv** [comment]
> I can get `currentExtent == (0,0)` and `minImageExtent == (1,1)`.

To be fair, that used to be true of NV's implementation too. I recently fixed the NV implementation to be spec conformant, and along the way the working group accepted changes to the spec that disambiguate spec language. (Previously, the spec was not self-consistent about what happens when a window has 0x0 dimensions, and it was previously ill-defined. Note that a VkImage cannot have 0x0 extent.)

**Comment by ianelliottus** [comment]
> Which sorta implies minImageExtent == maxImageExtent == currentExtent

The intention is that those values are equal for Windows.  An exception was noted for the 0x0 situation.

I'm not certain for Wayland and have no idea for VI.  I suggest splitting this into stand-alone issues for those platforms (so the entire issue doesn't remain open waiting for Wayland and VI platform-knowledgable people to comment).

> Should minImageExtent be allowed to be greater than currentExtent? Restriction on this seemed to have been removed in 1.0.63.

An exception was noted.  An attempt was made to explain that during such a situation, the application may not be able to create a swapchain.  We could repeat that info, but I think I've heard a desire to not be repetitive.  Is that okay?

**Comment by krOoze** [comment]
> The intention is that those values are equal for Windows. An exception was noted for the 0x0 situation.

Where is such an exception to this equality noted?

> An exception was noted. An attempt was made to explain that during such a situation, the application may not be able to create a swapchain.

I worked under such assumption in a Note in the newly merged branch. And clearly, swapchain could not be created in such a case.

But I don't see any exception noted anywhere saying that `minImageExtent` can be greater than `currentExtent`.

With the language saying must `minImageExtent`&le;`currentExtent` removed I guess it can be assumed for some platforms that may be the case. But still, that is a bit counter-intuitive and probably warrants a Note.

On Windows it would go against the language implying `minImageExtent == maxImageExtent == currentExtent` (can't be "greater than" if it is "equal"). But this language, AIS, doesn't feel strong enough (the equality is only implicitly assumed from all three variables "being the window size").

**Comment by ianelliottus** [comment]
> > The intention is that those values are equal for Windows. An exception was noted for the 0x0 situation.
> Where is such an exception to this equality noted?

The following:

> Note
On some platforms, vkGetPhysicalDeviceSurfaceCapabilitiesKHR can return in VkSurfaceCapabilitiesKHR a currentExtent that has a width and/or height equal to zero. This can happen, e.g., while a native window object is minimized. In this situation, the native window object may have to be restored from the minimized state before a VkSwapchainKHR can be created, because:
> 
> * The minImageExtent and maxImageExtent values returned by vkGetPhysicalDeviceSurfaceCapabilitiesKHR may match the currentExtent.
> 
> * On some platforms, the imageExtent member of VkSwapchainCreateInfoKHR must match the currentExtent returned by vkGetPhysicalDeviceSurfaceCapabilitiesKHR.

This text was part of the same change that removed the wording you talked about.  Looking at the internal discussion, the wording was removed because of the 0x0 case, in which case currentExtent will be 0x0 and not equal to {min|max}ImageExtent.

What change would you like?  Note: I can personally only address this for Windows and Linux because of my lack of specific knowledge about Wayland and VI.

**Comment by krOoze** [comment]
@ianelliottus If anything, that note enforces the notion they should always be equal:

> **may** match the currentExtent

I think it can be considered outdated, because of the accepted PR, where the Note says:

> The `vkGetPhysicalDeviceSurfaceCapabilitiesKHR` **may** return in `VkSurfaceCapabilitiesKHR` a `currentExtent` that has the `width` and/or `height` equal to 0, or that is less than its counterpart in `minImageExtent`.

It concedes that possibly `currentExtent`< `minImageExtent`. Though it is **not** setting an exception to restrictions raised elsewhere (such as the equality implied in platform-dependent surface chapters).

> What change would you like?

**If** there is indeed an exception intended here, it needs to say so at the same place the equality is being required.

So maybe something like:

> With Win32, `minImageExtent`, `maxImageExtent`, and `currentExtent` must: each equal to the window size. Except if `currentExtent` `width` or `height` is 0, then `minImageExtent` and/or `maxImageExtent` may: differ from `currentExtent`.

**Comment by jjulianoatnv** [comment]
The suggested language tweak from @krOoze makes sense. But maybe we should go farther.

I remember asking once if language stating that currentExtent must be between minImageExtent and maxImageExtent should be eliminated. I see the values as meaning this:
* The minImageExtent and maxImageExtent communicate the range of valid swapchain extents. It is not possible to create a swapchain with imageExtent that is smaller or larger than this range.
* The currentImageExtent communicates the current extent of the native window object.

If the currentImageExtent is smaller than minImageExtent, or is larger than maxImageExtent, then it is not possible to create a swapchain with imageExtent that matches that of the native window object, unless the native window object is resized.

Furthermore, when interacting with a windowing system that requires the swapchain imageExtent to match currentExtent, then it is not even possible to create a swapchain, unless the native window object is resized.

Note that such a rule set would address both the case where the native window object is "too small", and the case where the window is "too big". While I'm not aware of a Vulkan implementation that has a maxImageExtent which is smaller than the maximum possible extent of a native window object, this rule set could gracefully handle the scenario should it ever exist.

@krOoze, would such a rule set satisfy the issue you are raising?

**Comment by krOoze** [comment]
> If the currentImageExtent is smaller than minImageExtent, or is larger than maxImageExtent, then it is not possible to create a swapchain with imageExtent that matches that of the native window object, unless the native window object is resized.

I assume that's only specific to the platforms with the equality. Other platforms may still choose between `min` and `max`, even if `currentExtent` is not between those values.

> The minImageExtent and maxImageExtent communicate the range of valid swapchain extents. It is not possible to create a swapchain with imageExtent that is smaller or larger than this range.

I am not sure that is entirely accurate.
I don't think the swapchain itself cares that much about sizes (except the case of 0x0, which is actually `VkImage` requirement and a separate VU).
I think it communicates limits of the platform windowing system.

To me it really seems to communicate two things. [`minImageExtent`, `maxImageExtent`] 2D-range that is potentially `VK_SUBOPTIMAL`. And a single 2D-value (if valid) of `currentExtent`, which communicates the optimal size. In that interpretation it is reasonable to preserve some sanity relationships between `minImageExtent`, `currentExtent` and `maxImageExtent.

Then again I don't care that much as long as the restrictions are conceptually made.
From esthetical POV, I probably want to write something like (for every and all platforms):
```
if( caps.minImageExtent < caps.maxImageExtent ) // overloaded operator<, means w < w && h < h
    const bool hasCurrentExtent =
        !(caps.currentExtent.w == UINT32_MAX && caps.currentExtent.h == UINT32_MAX);
    const VkExtent2D swapchainSize =
        hasCurrentExtent ? caps.currentExtent : chooseIn(caps.minImageExtent, caps.maxImageExtent);
    return createSwapchain( swapchainSize );
else{
   throw SwapchainUncreatable; // probably pauses rendering somewhere
}
```

**Comment by jekstrand** [comment]
> I'm not certain for Wayland and have no idea for VI. I suggest splitting this into stand-alone issues for those platforms (so the entire issue doesn't remain open waiting for Wayland and VI platform-knowledgable people to comment).

For Wayland, the window system has no concept of a "current extent".  You simply present whatever size image you want and the window changes size to match.  The minimum and maximum extents are set based on driver maximum image sizes.

**Comment by ianelliottus** [comment]
I spent some time to look at the spec language, how it was changed for an internal MR, what problem that MR was trying to address (i.e. that windows can sometimes have a (0,0) size), and how it relates to this issue.  I believe that we need to change the spec, in order to be self-consistent and to better address the (0,0) case.  There's a grey area, that I will outline along with some alternatives.

The language of the spec for Microsoft Windows, and for X11 (both via Xlib and Xcb) is:

> minImageExtent, maxImageExtent, and currentExtent are the window size. Therefore, a swapchain’s imageExtent must match the window’s size.

The intention, for Microsoft Windows and X11, is exactly as @krOoze suggests:

> implies minImageExtent == maxImageExtent == currentExtent.

However, we hadn't considered the case of a minimized window, where (at least on Windows), the size becomes (0,0).  The recent change tried to address this (e.g. new VU that says VkSwapchainCreateInfoKHR::imageExtent must be non-zero, adds a note talking about (0,0)), but comes short of being self-consistent.

Quick note: I am not in a position to test this on Windows or Linux, with a real driver.  I'll assume that reports are true that minimized windows on MS Windows have a currentExtent of (0,0).  I'll also assume that this is true for Linux/X11 as well.  Again, I can't test it.  Given those assumptions ...

- In all (i.e. platform-independent) cases when currentExtent is (0,0), a swapchain can't be created (VkImage can't be (0,0) and the new VU states this for swapchains).  This will never happen in the case of platforms like Wayland, which always have the "special value" of currentExtent.
- For X11/Windows cases where currentExtent is NOT (0,0), we'd like minImageExtent == maxImageExtent == currentExtent.
- For X11/Windows cases where currentExtent IS (0,0), we have a choice about minImageExtent and maxImageExtent:
  1. We can say that minImageExtent == maxImageExtent == currentExtent (i.e. are (0,0)), and rely on the new VU to prevent swapchains being created with (0,0) images.
  2. We can say that minImageExtent and maxImageExtent are undefined.
  3. Something else, though I can't figure out a sensible value given that we want minImageExtent == maxImageExtent == currentExtent for non-(0,0) cases.

I am personally flexible on whether we got with alternative 1 or 2.  Alternative 1 problem means the smallest change to the spec (e.g. add a new platform-specific sentence to explain about the (0,0) case, including that a swapchain can't be created).  Others' thoughts?

I am willing to write the new spec language.

**Comment by krOoze** [comment]
I am kinda fan of making it `minImageExtent` always > (1, 1), and `maxImageExtent` == `currentExtent` == (0, 0). It has these nice properties:

1) `minImageExtent` is combination of all the restrictions. I.e. =max(`currentExtent` , `vkCreateSwapchain` restrictions) .
2) `maxImageExtent` is also a combination of all the restrictions. I.e. min(`currentExtent` , `vkCreateSwapchain` restrictions)
3) If (`maxImageExtent` < `minImageExtent`) then swapchain cannot be created.

It is possibly too complicated vs (i) if (0,0) is really truly meant to be the single and only exception where swapchain cannot be created.

BTW, If I think about it, it is bit weird. Apparently you [can paint minimized windows](https://msdn.microsoft.com/en-us/library/dd162483(v=vs.85).aspx), so maybe it should not even change size if minimized (i.e. comparable to when it is fully obscured by another window).

**Comment by ianelliottus** [comment]
> I am kinda fan of making it minImageExtent always > (1, 1)

Did you mean ">="?  I don't like that because what would the driver come up with?  This is a situation where a swapchain can't be created anyway (as communicated by a current Windows driver person).  It would also require modifying the generic (i.e. non-platform-specific) description for maxImageExtent, which says it's >= minImageExtent and currentExtent.

> It is possibly too complicated vs (i) if (0,0) is really truly meant to be the single and only exception where swapchain cannot be created

That's my understanding for Windows (again, by a current Windows driver person).

**Comment by nwnk** [comment]
> Quick note: I am not in a position to test this on Windows or Linux, with a real driver. I'll assume that reports are true that minimized windows on MS Windows have a currentExtent of (0,0). I'll also assume that this is true for Linux/X11 as well. Again, I can't test it. Given those assumptions ...

The minimum X11 window size is 1x1.

**Comment by ianelliottus** [comment]
Here's some proposed language, for both the WIN32, XLIB, and XCB extensions, in context with the current language for WIN32:

> With Win32, pname:minImageExtent, pname:maxImageExtent, and pname:currentExtent are the window size. Therefore, a swapchain's pname:imageExtent must: match the window's size.
> 
> When the window is minimized (and perhaps for other reasons), the pname:currentExtent may: have a pname:width and pname:height of 0. In such a case, it is not valid to create a swapchain, since the images of the swapchain must have non-zero extents. Additionally in such a case, pname:minImageExtent and pname:maxImageExtent will also have a pname:width and pname:height of 0, resulting in those values not being (respectively) the smallest and largest valid swapchain extents for the surface.

The first paragaph is current spec language (using asciidoc markup).  The 2nd paragraph is new.  Feedback?

**Comment by krOoze** [comment]
@ianelliottus 
1) undecorated "must"
2) resulting in -> causing
3) not being -> not represent
4) Not sure if ", resulting in [...]" is really needed. And technically it **is** still a restriction. You are still not allowed to create swapchain larger than `maxImageExtent`.
5) Might require some wholistic changes elsewhere, namely:
  a) The restriction on `minImageExtent` that was removed should be reinstated
  b) The note below `VkSwapchainCreateInfoKHR` would now be too general and should be simplified

**Comment by ianelliottus** [comment]
Thanks for the feedback @krOoze.  I have tried to address it.  Here's updated the language for the 2nd paragraph:

> When the window is minimized (and perhaps for other reasons), the pname:currentExtent may: have a pname:width and pname:height of 0.  In such a case, it is not valid to create a swapchain, since the images of the swapchain must: have non-zero extents.  Additionally in such a case, pname:minImageExtent and pname:maxImageExtent will also have a pname:width and pname:height of 0, causing those values to not represent (respectively) the smallest and largest valid swapchain extents for the surface.

I think that addresses all of your feedback on that paragraph.

> The restriction on minImageExtent that was removed should be reinstated

I agree.  Done.

> The note below VkSwapchainCreateInfoKHR would now be too general and should be simplified using the clarifications you are making

I agree it can be simplified.  It seemed like mentioning the "minimized" case was good to keep, but wonder if it's still too wordy.  Here's what I have at this point:

> During some situations, sname:VkSwapchainKHR cannot: be created because of the Valid Usage of the pname:imageExtent parameter.  For example, a minimized window on some platforms may: result in
flink:vkGetPhysicalDeviceSurfaceCapabilitiesKHR returning in slink:VkSurfaceCapabilitiesKHR a pname:currentExtent, pname:minImageExtent, and pname:maxImageExtent that have the pname:width and/or pname:height equal to 0.

Additional feedback?

**Comment by krOoze** [comment]
Yeah. When I was making that Note, I assumed there might be more cases that cause uncreatable swapchain, therefore the chosen sentence structure. Maybe now it could just be something like:

> On some platforms it is normal that `maxImageExtent` **may** become `(0, 0)`, for example when the window is minimized. In such case it is not possible to create swapchain due to the Valid Usage requirements.

The platform specific part could perhaps also be shortened. And maybe clarified at the same time. E.g.:

> For Win32 surface, `minImageExtent`, `maxImageExtent`, and `currentExtent` values returned by `vkGetPhysicalDeviceSurfaceCapabilitiesKHR` **must** always equal.
>
>For Win32 surface, `currentExtent` **must** have both `width` and `height` greater than 0, or both of them 0.
>
>Note:
>> Due to above restrictions, on Win32 it is possible to create new swapchain only with the current size of the window.
>>
>>The window size **may** become `(0, 0)` on this platform, for example when window is minimized, and so swapchain cannot: be created until the size changes.

The Note is optional. Depends how much we want to insult the intelligence of the readers...

**Comment by oddhack** [comment]
This should be fixed in the 1.0.67 spec update.
